File: rust/day-04/src/main.rs
fn main() {
    let input = include_str!("../input.txt");
    let mut input = input.split("\n").collect::<Vec<&str>>();
    input.pop();
    part1(&input);
    part2(&input);
}

fn line_to_vec(line: &str) -> Vec<&str> {
    let split: Vec<&str> = line.split(":").collect();

    split[1].split(" | ").collect()
}

fn vec_to_count(numbers: &Vec<&str>) -> usize {
    let winning: Vec<&str> = numbers[0].split(" ").collect();
    let nums: Vec<&str> = numbers[1].split(" ").collect();
    
    nums.iter().filter(|n| !n.is_empty() && winning.contains(n)).count()
}

fn count_to_score(amount: usize) -> usize {
    if amount == 0 {
        return 0;
    }

    2_usize.pow(amount as u32 - 1) 
}

fn part1(input: &Vec<&str>) {
    let sum: usize = input.iter()
        .map(|l| line_to_vec(l))
        .map(|l| count_to_score(vec_to_count(&l)))
        .sum();

    println!("{}", sum);
}

fn part2(input: &Vec<&str>) {
    let mut cards: Vec<u32> = vec![1; input.len()];

    let amounts: Vec<usize> = input.iter()
        .map(|l| line_to_vec(l))
        .map(|l| vec_to_count(&l)).collect();

    for (i, amount) in amounts.iter().enumerate() {
        if amount == &0 {
            continue;
        }

        let before = cards.clone();
        for j in i+1..i+amount+1 {
            cards[j] += before[i];
        } 
    }

    let sum: u32 = cards.iter().sum();
    println!("{}", sum);
}

File: elixir/day-04/main.exs
defmodule Day4 do

  defp get_winning_and_cards(line) do
    [_num, cards] = String.split(line, ": ")
    String.split(cards, " | ")
  end

  defp get_card_numbers(cards) do
    String.split(cards, " ") |> Enum.filter(fn s -> s !== "" end)
  end

  defp get_score_from_count(count) do
    case count do
      0 -> 0
      _ -> Integer.pow(2, count - 1)
    end
  end

  defp update_card_counts(cards, amount, first, last) do
    cond do
      first === last -> cards
      true -> 
        cards = List.update_at(cards, first, &(&1 + amount))
        update_card_counts(cards, amount, first + 1, last)
    end 
  end

  defp get_card_counts(counts, cards, _current) when length(counts) === 1 do
    cards
  end
  
  defp get_card_counts(counts, cards, current) do
    [head | tail] = counts
    amount = Enum.at(cards, current - 1)
    cards = update_card_counts(cards, amount, current, current + head)
    get_card_counts(tail, cards, current + 1)
  end

  defp get_count(line) do
    [winning, cards] = get_winning_and_cards(line)
    winning = get_card_numbers(winning)
    cards = get_card_numbers(cards)
    Enum.count(cards, fn c -> Enum.member?(winning, c) end)
  end

  defp get_score(line) do
    count = get_count(line)
    get_score_from_count(count)
  end

  def part1(input) do
    score = Enum.map(input, fn s -> get_score(s) end) |> Enum.sum()
    IO.inspect(score)
  end

  def part2(input) do
    matching_counts = Enum.map(input, fn s -> get_count(s) end)
    total_counts = List.duplicate(1, length(input))
    final = get_card_counts(matching_counts, total_counts, 1)
    sum = Enum.sum(final)
    IO.inspect(sum)
  end
end

file = File.read!("./input.txt")
file = String.trim(file)
file = String.split(file, "\n")

Day4.part1(file)
Day4.part2(file)
